src/extensions/data_array.js

```diff
@@ -67,24 +96,101 @@ export default () => {
       });
     };
 
+    const createInstances = observationObjects => {
+      return new Promise(resolve => {
+        db().then(models => {
+          Promise.all(observationObjects.map(observationObj => {
+            const newReq = Object.create(req);
+            newReq.body = observationObj.body;
+            return models.createInstance(observations, newReq, exclude).then(
+              instance => {
+              observationObj.instanceId = instance.id;
+              return Promise.resolve();
```

Comment:
After testing the approach you suggested (`return Promise.resolve(instance.id)` instead of assigning to `observationObj`), I found it works as you expected -- I thought I had tried that but it seems my initial approach was somehow different. I think I didn't realize the magic of using `map` in conjunction with `Promise.all` until later and by then I had mixed a couple of different approaches. In any event, the code is very efficient now and does what it is supposed to do.